keps/sig-apps/sidecarcontainers.md

```diff
@@ -154,8 +155,34 @@ The proposal can broken down into four key pieces of implementation that all rel
 * Sidecars are terminated after normal containers
 * Sidecars start before normal containers
 
+#### API Changes:
 As this is a change to the Container spec we will be using feature gating, you will be required to explicitly enable this feature on the api server as recommended [here](https://github.com/kubernetes/community/blob/master/contributors/devel/api_changes.md#adding-unstable-features-to-stable-versions).
 
+New field `Type` will be added to the lifecycle struct:
+
+```go
+type Lifecycle struct {
+  // Type
+  // One of Standard, Sidecar.
+  // Defaults to Standard
+  // +optional
+  Type LifecycleType `json:"type,omitempty" protobuf:"bytes,3,opt,name=type,casttype=LifecycleType"`
+}
+```
+New type `LifecycleType` will be added with two constants:
+```go
+// LifecycleType describes the lifecycle behaviour of the container
+type LifecycleType string
+
+const (
+  // LifecycleTypeStandard is the default container lifecycle behaviour
+  LifecycleTypeStandard LifecycleType = "Standard"
+  // LifecycleTypeStandard means that the container will start up before standard containers and be terminated after
+  LifecycleTypeSidecar LifecycleType = "Sidecar"
+)
+```
+Note that currently the `lifecycle` struct is only used for `preStop` and `postStop` so we will need to change its description to reflect the expansion of its uses.
+
 #### Kubelet Changes:
```

Comment:
Porting back several discussions about addressing restart policy and pod phase in another thread, since this is the main one to discuss design: https://github.com/kubernetes/enhancements/issues/753#issuecomment-488096569

We have been trying side car in production internally and according to various failure lessons we learned about sidecar, we found that having the following would be useful in production:

1. Pod phase to be only calculated based on "Standard" containers as this is where real meaningful work is done, and also to avoid unnecessary retries especially for run-to-finish workload such as Job/CronJob etc.

2. Restart policy to be only applied to "Standard" containers and to always restart "Sidecar" if "Standard" is not finished, aka running or needs to be (re)started, to cope with the "baby-sitting" role of sidecar and a historical assumption from "Standard" containers that sidecars might be temporarily unavailable

Reply:
Hi @zhan849, can we please keep this pr about the API implementation. API reviewers have limited time and I don't want this to be any more complicated to review than it is

Reply:
Sure we can include additional behavior tunings in another KEP if the scope of this KEP is already big.